RHDEVDOCS-4249 - CMO config settings for topology spread constraints (#52118)
openshift/openshift-docs: modules/monitoring-setting-up-pod-topology-spread-constraints-for-prometheus.adoc

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: PROCEDURE
[id="setting-up-pod-topology-spread-constraints-for-prometheus_{context}"]
= Setting up pod topology spread constraints for Prometheus

> For core {product-title} platform monitoring, you can set up pod topology spread constraints for Prometheus to fine tune how pod replicas are scheduled across zones.

Author: changed

> Doing so helps ensure that Prometheus pods are highly available and run more efficiently because workloads are spread across different data centers or hierarchical infrastructure zones.

Author: added

> Doing so helps ensure that Prometheus pods are highly available and run more efficiently because workloads are spread across nodes in different data centers or hierarchical infrastructure zones.

Reviewer: Doing so helps ensure
->
Doing so ensures

Author: added

> You configure pod topology spread constraints for core platform monitoring in the `cluster-monitoring-config` config map.

Reviewer: You configure
->
You can configure

Author: changed

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* You have access to the cluster as a user with the `cluster-admin` role.
* You have created the `cluster-monitoring-config` `ConfigMap` object.

.Procedure

. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` namespace:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

> . Add  values for the following settings under `data/config.yaml/prometheusK8s` and `data/config.yaml/alertmanagermain` to configure topology spread constraints:

Reviewer 2: If we have a dedicated section for Alertmanager
```suggestion
. Add  values for the following settings under `data/config.yaml/prometheusK8s` to configure topology spread constraints:
```

Author: fixed

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      topologySpreadConstraints:
      - maxSkew: 1 <1>
        topologyKey: monitoring <2>
        whenUnsatisfiable: DoNotSchedule <3>
        labelSelector: 
          matchLabels: <4>

>             prometheus: k8s

Reviewer 2: ```suggestion
            app.kubernetes.io/name: prometheus
```

Author: fixed

----

> <1> Specify a numeric value for `maxSkew`, which defines the degree to which pods can be unevenly distributed. 

Reviewer: `can be` changed to `may be` is better
https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#spread-constraint-definition

Author: Changed to "might be". The IBM Style Guide does not allow us to use "may" when we mean "might."

This field is required, and the value must be greater than zero. 
The value specified has a different effect depending on what value you specify for `whenUnsatisfiable`.

> <2> Specify a node label for `topologyKey`.

Reviewer 2: ```suggestion
<2> Specify a key of node labels for `topologyKey`.
```

Author: changed

This field is required.

> Nodes that match this label are considered to be in the same domain, and. 
> the pod scheduler tries to balance the number of pods in each domain.

Reviewer 2: ```suggestion
Nodes that have a label with this key and identical values are considered to be in the same topology.
The scheduler will try to put a balanced number of pods into each domain.
```

Author: changed

> The value `non-exist` indicates that a node label does not exist and that therefore the pods can be scheduled to any node, whether or not the node has the `monitoring` label.

Reviewer 2: I didn't find a reference to this `non-exist` value, do you have a pointer where I can read more info on it?

Author: This was mentioned in the text case in Polarion at https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-47393

Reviewer: `non-exist` in the case just mean the topologyKey does not exist, you can replace to any value, that's does not mean you must set `non-exist`

Author: Ok. Thanks! I'll remove this.

<3> Specify a value for `whenUnsatisfiable`.
This field is required.
Available options are `DoNotSchedule` and `ScheduleAnyway`.
Specify `DoNotSchedule` if you want the `maxSkew` value to define the maximum difference allowed between the number of matching pods in the target topology and the global minimum. 
Specify `ScheduleAnyway` if you want the scheduler to still schedule the pod but to give higher priority to nodes that might reduce the skew.
<4> Specify a value for `matchLabels`. This value is used to identify the set of matching pods to which to apply the constraints.

. Save the file to apply the changes automatically.
+
[WARNING]
====

> When you save changes to a `cluster-monitoring-config` config map, the pods and other resources in the `openshift-monitoring` project might be redeployed. 

Reviewer: When you save changes to the `cluster-monitoring-config` config map

Author: fixed
